File: scripts/utf8_structured_write.py
#!/usr/bin/env python3
# ---------------------------------------------------------------------
# lupopedia.headers:
#   header_format_version: "4.1.3"
#   file_path_from_root: "scripts/utf8_structured_write.py"
#   web_path: "https://www.lupopedia.com/lupopedia/scripts/utf8_structured_write.py"
#   status: "active"
#   when_updated: "20260418201851"
#   trust_tier: "canonical"
#   questions_toon: null
#   memory_toon: "memory/development/canonical/1026/04/utf8-structured-write.toon"
#   atoms_toon: null
#   transcript_jsonl: "0/development/utf8-structured-write"
#   artifact_type: implementation
#   artifact_kind: tool
#   channel_key: "development"
#   federation_node_id: 0
#   thread_id: ""
#   content_id: null
#   content_parent_id: 16
#   content_slug: "utf8-structured-write"
#   default_collection_id: null
#   lupopedia.schema: implementation
#   title: "UTF-8 structured write helper for JSON and TOON outputs"
#   summary: "normalize_known_mojibake, prepare_for_filesystem_write, dumps_json_ready_for_write, write_text_utf8; forward-only; aligned with PHP Utf8StructuredWrite."
# ---------------------------------------------------------------------
"""
Forward-only UTF-8 helpers for JSON / TOON-shaped text before disk write.

Manual fix for a single corrupted file: copy through Windows Notepad (UTF-8),
per operator runbook. Do not add one-off encoding scripts unless systemic.

Import from generators:
    from utf8_structured_write import (
        prepare_for_filesystem_write,
        dumps_json_ready_for_write,
        write_text_utf8,
    )
"""

import json
from pathlib import Path
from typing import Any, Dict, Tuple

# Same three replacements as PHP Utf8StructuredWrite::mojibakeReplacementPairs() (byte-identical UTF-8 runs).
_MOJIBAKE_PAIRS: Tuple[Tuple[str, str], ...] = (
    # Corrupted: smart double-quote (U+201D class) mis-decoded as UTF-8 mojibake. Target: ASCII " (U+0022).
    ("\xc3\xa2\xe2\x82\xac\xc2\x9d", '"'),
    # Corrupted: dash-like mojibake where hyphen-minus was intended. Target: ASCII hyphen-minus.
    ("\xc3\xa2\xe2\x82\xac\xe2\x80\x98", "-"),
    # Corrupted: ellipsis (U+2026 class) mis-decoded chain. Target: remove (empty string).
    ("\xc3\xa2\xc5\x93\xe2\x80\xa6", ""),
)


def normalize_known_mojibake(text: str) -> str:
    if not text:
        return text
    out = text
    for bad, good in _MOJIBAKE_PAIRS:
        out = out.replace(bad, good)
    return out


def coerce_to_utf8(text: str) -> str:
    """
    Apply the small known-mojibake table, then normalize to a UTF-8-round-trippable str.

    This is **not** byte-level recovery equivalent to PHP ``iconv('UTF-8', 'UTF-8//IGNORE', ...)`` on
    arbitrary bytes: callers must pass ``str``. The encode/decode step uses ``errors="replace"`` so
    lone surrogate code points (invalid in UTF-8) become U+FFFD on the round-trip; typical JSON text
    never hits that path. For arbitrary binary repair, use a dedicated binary pipeline, not this helper.
    """
    out = normalize_known_mojibake(text)
    return out.encode("utf-8", errors="replace").decode("utf-8")


def prepare_for_filesystem_write(text: str, file_path: str = "") -> Dict[str, Any]:
    """
    Validate / normalize ``text`` for safe UTF-8 persistence as a single blob.

    * ``file_path``: pass-through only (symmetry with PHP ``prepareForFilesystemWrite`` and for
      outer-layer logging context). Intentionally unused inside this function.
    * **JSONL / framing:** this function does **not** define JSONL line boundaries or trailing
      newline policy beyond whatever ``text`` already contains; it only checks the string can be
      encoded as UTF-8 after mojibake repair. Callers own JSONL assembly (one write per line, etc.).
    """
    if not isinstance(text, str):
        return {"ok": False, "text": "", "reason": "not_string", "changed": False}
    before = text
    after = coerce_to_utf8(before)
    try:
        after.encode("utf-8")
    except UnicodeEncodeError:
        return {
            "ok": False,
            "text": "",
            "reason": "invalid_utf8_after_coerce",
            "changed": before != after,
        }
    return {
        "ok": True,
        "text": after,
        "reason": "",
        "changed": before != after,
    }


def write_text_utf8(path: Path, text: str, *, dry_run: bool = False) -> bool:
    prep = prepare_for_filesystem_write(text, str(path))
    if not prep.get("ok"):
        return False
    if dry_run:
        return True
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(prep["text"], encoding="utf-8", newline="\n")
    return True


def dumps_json_ready_for_write(obj: Any, **kwargs: Any) -> str:
    """
    ``json.dumps`` the object, ensure exactly one trailing newline, then UTF-8 prepare.

    The trailing newline is intentional so callers can append or write one JSON record per line
    without adding separators here. JSONL consumers still split on newlines themselves.
    """
    body = json.dumps(obj, ensure_ascii=False, **kwargs)
    if not body.endswith("\n"):
        body += "\n"
    prep = prepare_for_filesystem_write(body, "")
    if not prep.get("ok"):
        raise ValueError("structured UTF-8 prepare failed: %r" % prep.get("reason"))
    return prep["text"]

File: archive/lupo-includes/classes/Utf8StructuredWrite.php
<?php
# ---------------------------------------------------------------------
# lupopedia.headers:
#   header_format_version: "4.1.3"
#   file_path_from_root: "lupo-includes/classes/Utf8StructuredWrite.php"
#   web_path: "https://www.lupopedia.com/lupopedia/lupo-includes/classes/Utf8StructuredWrite.php"
#   status: "active"
#   when_updated: "20260418200841"
#   trust_tier: "canonical"
#   questions_toon: null
#   memory_toon: "lupo-memory/development/canonical/1026/04/utf8structuredwrite.toon"
#   atoms_toon: null
#   transcript_jsonl: "0/development/utf8structuredwrite"
#   artifact_type: implementation
#   artifact_kind: tool
#   channel_key: "development"
#   federation_node_id: 0
#   thread_id: ""
#   content_id: null
#   content_parent_id: 16
#   content_slug: "utf8structuredwrite"
#   default_collection_id: null
#   lupopedia.schema: implementation
#   title: "Utf8StructuredWrite structured UTF-8 write guard"
#   summary: "Forward-only UTF-8 guard for JSON, JSONL, and TOON-shaped text before filesystem write; known mojibake repair; invalid UTF-8 blocks write."
# ---------------------------------------------------------------------
/**
 * Forward-only UTF-8 guard for structured artifacts (JSON, JSONL, TOON-shaped text).
 *
 * Doctrine: normalize known UTF-8/cp1252 mojibake sequences before write; verify valid UTF-8;
 * block write and return failure if still invalid. Does not scan the repo or rewrite existing files.
 *
 * Policy: this class enforces UTF-8 byte safety and a tiny explicit mojibake repair table only.
 * It does not parse or validate JSON, JSONL, TOON, or other semantic structure; callers must
 * validate meaning (schema, line shape, channel rules) separately.
 *
 * Manual fix for isolated hand-edited files: copy through Windows Notepad (UTF-8) per operator runbook.
 */
final class Utf8StructuredWrite
{
    /**
     * Known bad byte runs (double-UTF-8 or common cp1252-misread artifacts) replaced in order.
     *
     * @return array<string,string>
     */
    private static function mojibakeReplacementPairs()
    {
        return array(
            // Broken: UTF-8 mojibake from smart double-quote (U+201D style misread). Target: ASCII " (U+0022).
            "\xC3\xA2\xE2\x82\xAC\xC2\x9D" => '"',
            // Broken: UTF-8 mojibake run often seen where ASCII hyphen-minus was intended. Target: '-'.
            "\xC3\xA2\xE2\x82\xAC\xE2\x80\x98" => '-',
            // Broken: UTF-8 mojibake from ellipsis (U+2026 style misread). Target: drop (empty string).
            "\xC3\xA2\xC5\x93\xE2\x80\xA6" => '',
        );
    }

    /**
     * Replace known mojibake substrings (UTF-8 byte runs).
     *
     * @param string $text
     * @return string
     */
    public static function normalizeKnownMojibake($text)
    {
        if (!is_string($text) || $text === '') {
            return is_string($text) ? $text : '';
        }
        $pairs = self::mojibakeReplacementPairs();
        return str_replace(array_keys($pairs), array_values($pairs), $text);
    }

    /**
     * True if PHP considers the string valid UTF-8.
     *
     * @param string $text
     * @return bool
     */
    public static function isValidUtf8($text)
    {
        if (!is_string($text)) {
            return false;
        }
        if ($text === '') {
            return true;
        }
        if (function_exists('mb_check_encoding')) {
            return (bool) mb_check_encoding($text, 'UTF-8');
        }
        $r = @preg_match('//u', $text);
        return $r === 1;
    }

    /**
     * Normalize mojibake, then strip invalid UTF-8 bytes if iconv is available.
     *
     * @param string $text
     * @return string
     */
    public static function coerceToUtf8($text)
    {
        if (!is_string($text)) {
            return '';
        }
        $text = self::normalizeKnownMojibake($text);
        if (!self::isValidUtf8($text) && function_exists('iconv')) {
            $fixed = @iconv('UTF-8', 'UTF-8//IGNORE', $text);
            if (is_string($fixed)) {
                $text = $fixed;
            }
        }
        return $text;
    }

    /**
     * Prepare one structured text blob (e.g. a JSONL line) for filesystem write.
     *
     * @param string $text Already-encoded JSON/JSONL line (no trailing newline required).
     * @param string $file_path Pass-through context for outer layers (e.g. target path for logging on failure).
     *        Intentionally unused inside this method; kept so callers like HermesService can pass a path
     *        without this class opening or touching the filesystem.
     * @return array { ok: bool, text: string, reason: string, changed: bool }
     */
    public static function prepareForFilesystemWrite($text, $file_path = '')
    {
        if (!is_string($text)) {
            return array(
                'ok' => false,
                'text' => '',
                'reason' => 'not_string',
                'changed' => false,
            );
        }
        $before = $text;
        $after = self::coerceToUtf8($text);
        $changed = ($after !== $before);
        if (!self::isValidUtf8($after)) {
            return array(
                'ok' => false,
                'text' => '',
                'reason' => 'invalid_utf8_after_coerce',
                'changed' => $changed,
            );
        }
        return array(
            'ok' => true,
            'text' => $after,
            'reason' => '',
            'changed' => $changed,
        );
    }
}
